src/nexus/writer.py

```diff
@@ -57,7 +70,10 @@ def clean(self, s):
 
     @property
     def characters(self):
-        return list(self.data.keys())
+        if self._characters is None:
+            self._characters = list(sorted(self.data.keys(), key=lambda x: self._chars_in.index(x)
+                                           if self.preserve_order else x))
+        return self._characters
```

Comment:
I'm not sure caching this is a good idea -- what if data changes?

Reply:
I agree (and the same goes for `taxa`). Since `NexusWriter` has an explicit `add` method, we cannot simply assume that `taxa` or `characters` will only be evaluated after the last `add` call.

Reply:
This could be solved by setting `self._characters` and `self._taxa` to `None` after methods `remove()`, `remove_taxon()`, `remove_character()`, and add().